llvm/test/Feature/Repo/repo_pruning_llvm.global_ctors.ll

```diff
@@ -0,0 +1,55 @@
+; This testcase tests the program repository pruning pass for the llvm.global_ctors.
+;
+; The testcase includes three steps:
+; Step 1: Build the code and create the database 'clang.db' which contains all Tickets.
+; Step 2: Dump all compilation members in the database;
+; Step 3: Re-build the same IR code again.
+; Step 4: Dump all compilation members in the database again;
+; Step 5: Check that the step 2 and step 4 generate the same compilation members.
+
+; RUN: rm -f %t.db
+; RUN: env REPOFILE=%t.db opt -O3 -S -mtriple=x86_64-pc-linux-gnu-repo %s -o %t.ll
+; RUN: env REPOFILE=%t.db llc -mtriple="x86_64-pc-linux-gnu-repo" -filetype=obj %t.ll -o %t.o
+; RUN: env REPOFILE=%t.db pstore-dump  -all-compilations %t.db > %t.log
+; RUN: env REPOFILE=%t.db opt -O3 -S -mtriple=x86_64-pc-linux-gnu-repo %s -o %t1.ll
+; RUN: env REPOFILE=%t.db llc -mtriple="x86_64-pc-linux-gnu-repo" -filetype=obj %t1.ll -o %t1.o
+; RUN: env REPOFILE=%t.db pstore-dump  -all-compilations %t.db > %t1.log
+; RUN: diff %t.log %t1.log
+
+target triple = "x86_64-pc-linux-gnu-elf"
+
+%"String" = type { i8* }
+%"Entry" = type { %"String" }
+%"Array" = type { %"Entry"*, i64 }
+
+@_Names = global [1 x %"Entry"] zeroinitializer, align 16
+@.str = private unnamed_addr constant [5 x i8] c"Read\00", align 1
+@llvm.global_ctors = appending global [1 x { i32, void ()*, i8* }] [{ i32, void ()*, i8* } { i32 65535, void ()* @_GLOBAL__sub_I_EnumTables.cpp, i8* null }]
+
+declare i64 @strlen(i8*)
+define linkonce_odr hidden void @_A(%"String"*, i8*) align 2 {
```

Comment:
Can we simplify these functions? Does _A() need to be linkonce?

Reply:
> 
> 
> Can we simplify these functions? 
>
I can't remove any instructions in all these functions. When I tried to simply these functions, the issue isn't shown in the master branch.

>
>Does _A() need to be linkonce?

No, remove the `linkonce_odr hidden` in the function definition.

Reply:
> I can't remove any instructions in all these functions. When I tried to simply these functions, the issue isn't shown in the master branch.

That's surprising. Why does that happen? What is "special" about these instructions that exposes the problem?

Reply:
> 
> 
> > I can't remove any instructions in all these functions. When I tried to simply these functions, the issue isn't shown in the master branch.
> 
> That's surprising. Why does that happen? What is "special" about these instructions that exposes the problem?
>
Good question. I will try to explain the reason.

In order to explore how to reproduce the problem, I printed IR code after all optimisation passes using the `-print-after-all` command. The following steps need to be satisfied to reproduce the issue. (The steps for the first-time build follow the order of executing the optimisation passes)
1. After the `globalopt` pass (first-time), both `_Names` and `llvm.global_ctors` keep the same. There is **NO** global variable optimisation.
```
@_Names = global [1 x %"Entry"] zeroinitializer, align 16
@llvm.global_ctors = appending global [1 x { i32, void ()*, i8* }] [{ i32, void ()*, i8* } { i32 65535, void ()* @_GLOBAL__sub_I_EnumTables.cpp, i8* null }]
```
2. After several optimisation passes, the functions `_A` and `_GLOBAL__sub_I_EnumTables.cpp` have been optimised to the following simple functions.
```
define void @_A(%String* nocapture, i8*) local_unnamed_addr #1 align 2 !repo_ticket !3 %3 = getelementptr inbounds %String, %String* %0, i64 0, i32 0
store i8* %1, i8** %3, align 8
ret void
}
…
define internal void @_GLOBAL__sub_I_EnumTables.cpp() #1 !repo_ticket !12 {
store i8* getelementptr inbounds ([5 x i8], [5 x i8]* @.str, i64 0, i64 0), i8** getelementptr inbounds ([1 x %Entry], [1 x %Entry]* @_Names, i64 0, i64 0, i32 0, i32 0), align 16
ret void
}
```
3. After the `elim-avail-extern` pass, there is **NO** change since there is **NO** `available_externally` linkage type for all global objects.
4. Since the functions `_A` and `_GLOBAL__sub_I_EnumTables.cpp` have been simplified in the step 2, both `_Names` and `llvm.global_ctors` could be optimised during the global variable optimisation. After the `globalopt` pass (second-time), `_Names` and `llvm.global_ctors` changed to:
```
@_Names = local_unnamed_addr global [1 x %Entry] [%Entry { %String { i8* getelementptr inbounds ([5 x i8], [5 x i8]* @.str, i64 0, i64 0) } }], align 16, !repo_ticket !0
@llvm.global_ctors = appending global [0 x { i32, void ()*, i8* }] zeroinitializer
```
If the functions `_A` and `_GLOBAL__sub_I_EnumTables.cpp` are too simple (such as mentioned in step 2), both `_Names` and `llvm.global_ctors` are optimised when running the `globalopt` pass in the step 1. When recompile the test with the existing database, both `_Names` and `llvm.global_ctors` are optimised when running the `globalopt` pass in the step 1. Therefore, the issue can’t be reproduced.

Reply:
Ok, that makes sense. Thank you.